src/core/modules/impl/CacheableStateEvaluator.ts

```diff
@@ -35,79 +34,61 @@ export class CacheableStateEvaluator extends DefaultStateEvaluator {
     currentTx: CurrentTx[]
   ): Promise<EvalStateResult<State>> {
-    const requestedBlockHeight = executionContext.blockHeight;
-    this.cLogger.debug(`Requested state block height: ${requestedBlockHeight}`);
-
     const cachedState = executionContext.cachedState;
-    if (cachedState?.cachedHeight === requestedBlockHeight) {
+    if (cachedState && cachedState.sortKey == executionContext.requestedSortKey) {
+      this.cLogger.info(
+        `Exact cache hit for sortKey ${executionContext?.contractDefinition?.txId}:${cachedState.sortKey}`
+      );
       executionContext.handler?.initState(cachedState.cachedValue.state);
       return cachedState.cachedValue;
     }
 
-    this.cLogger.debug('executionContext.sortedInteractions', executionContext.sortedInteractions.length);
-
-    const sortedInteractionsUpToBlock = executionContext.sortedInteractions.filter((tx) => {
-      return tx.node.block.height <= executionContext.blockHeight;
-    });
-
-    let missingInteractions = sortedInteractionsUpToBlock.slice();
+    const missingInteractions = executionContext.sortedInteractions;
 
-    this.cLogger.debug('missingInteractions', missingInteractions.length);
-
-    // if there was anything to cache...
-    if (sortedInteractionsUpToBlock.length > 0) {
-      if (cachedState != null) {
-        this.cLogger.debug(`Cached state for ${executionContext.contractDefinition.txId}`, {
-          cachedHeight: cachedState.cachedHeight,
-          requestedBlockHeight
-        });
-
-        // verify if for the requested block height there are any interactions
-        // with higher block height than latest value stored in cache - basically if there are any non-cached interactions.
-        missingInteractions = sortedInteractionsUpToBlock.filter(
-          ({ node }) => node.block.height > cachedState.cachedHeight && node.block.height <= requestedBlockHeight
-        );
-      }
-
-      this.cLogger.debug(`Interactions until [${requestedBlockHeight}]`, {
-        total: sortedInteractionsUpToBlock.length,
-        cached: sortedInteractionsUpToBlock.length - missingInteractions.length
-      });
-
-      // TODO: this is tricky part, needs proper description
-      // for now: it prevents from infinite loop calls between calls that are making
-      // internal interact writes.
-      for (const entry of currentTx || []) {
-        if (entry.contractTxId === executionContext.contractDefinition.txId) {
-          const index = missingInteractions.findIndex((tx) => tx.node.id === entry.interactionTxId);
-          if (index !== -1) {
-            this.cLogger.debug('Inf. Loop fix - removing interaction', {
-              height: missingInteractions[index].node.block.height,
-              contractTxId: entry.contractTxId,
-              interactionTxId: entry.interactionTxId
-            });
-            missingInteractions.splice(index);
-          }
+    // TODO: this is tricky part, needs proper description
```

Comment:
well, this is next thing to be refactored, but not in this PR...instead of keeping the track of called contracts/transactions in this weird array, maybe we could somehow use the fact, the we keep the whole Contracts tree (i.e. each contract that is being read by the "Smartweave.contracts.readContractState" has a "parent" contract set...